### ISSUE TITLE
Closing file handles on HUP

### ISSUE THREAD

**Issue by browning** [benoitc/gunicorn#1012]
Hi, 

I came across this issue from 2013 https://github.com/benoitc/gunicorn/issues/627  , the fix for it was made in the fix/627 branch but it looks like that branch was never merged into master.

I was wondering if that could be merged now? or if there was a reason it was not merged?

Thanks!

**Comment by benoitc** [comment]
will investigate that, thanks for the head up.

**Comment by parasyte** [comment]
I just noticed this issue on one of our machines in production. I recently added logrotate which sends HUP. The logrotate conf also compresses old logs, so we end up with the master process retaining an open FD to all of the deleted files. The result is that `df` and `du` disagree on how much space is available on the partition. Here are some diagnostics:

```
root@ip-172-16-98-67:/mnt# df -h | egrep 'Filesystem|/mnt'
Filesystem      Size  Used Avail Use% Mounted on
/dev/xvdb        74G   32G   39G  45% /mnt
```

```
root@ip-172-16-98-67:/mnt# du -a -h
879M    ./logs/pn-auth/pn-auth.log
28M     ./logs/pn-auth/pn-auth-error.log.1
53M     ./logs/pn-auth/pn-auth-error.log
178M    ./logs/pn-auth/pn-auth.log.1.gz
168M    ./logs/pn-auth/pn-auth.log.2.gz
175M    ./logs/pn-auth/pn-auth.log.3.gz
1.5G    ./logs/pn-auth
1.5G    ./logs
16K     ./lost+found
1.5G    .
```

:flushed: 

```
root@ip-172-16-98-67:/mnt# lsof | grep /mnt
gunicorn    865    pn-auth    3w      REG             202,16    29297709     622595 /mnt/logs/pn-auth/pn-auth-error.log.1
gunicorn    865    pn-auth    5w      REG             202,16 11316232586     622596 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   10w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   11w      REG             202,16 14908907332     622600 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   12w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   13w      REG             202,16  1183851759     622603 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   14u      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   15u      REG             202,16  1111137094     622606 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   16u      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   17u      REG             202,16  1132489097     622609 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   18w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   19w      REG             202,16 14908907332     622600 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   20w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   21w      REG             202,16  1183851759     622603 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   22w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   23w      REG             202,16  1104112244     622604 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   24u      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   25u      REG             202,16  1128364446     622608 /mnt/logs/pn-auth/pn-auth.log.1 (deleted)
gunicorn    865    pn-auth   26u      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn    865    pn-auth   27u      REG             202,16   896182612     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49392    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49392    pn-auth   27w      REG             202,16   896183415     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49395    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49395    pn-auth   27w      REG             202,16   896183415     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49397    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49397    pn-auth   27w      REG             202,16   896183415     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49398    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49398    pn-auth   27w      REG             202,16   896183415     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49399    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49399    pn-auth   27w      REG             202,16   896183415     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49400    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49400    pn-auth   27w      REG             202,16   896183415     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49401    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49401    pn-auth   27w      REG             202,16   896183752     622610 /mnt/logs/pn-auth/pn-auth.log
gunicorn  49402    pn-auth   26w      REG             202,16    54558380     622599 /mnt/logs/pn-auth/pn-auth-error.log
gunicorn  49402    pn-auth   27w      REG             202,16   896183752     622610 /mnt/logs/pn-auth/pn-auth.log
```

:confounded:

We are going to run out of disk space in a few days and have to kill the master to release these FDs.

**Comment by tilgovi** [comment]
Merged in ed9d5dd8247673babb4ce45a926575d883bc4500 and released in 19.4.
